rat/symbol_table.py

```diff
@@ -0,0 +1,270 @@
+from dataclasses import dataclass, field
+from enum import Enum
+import numpy
+import pandas
+import pprint
+from typing import Dict, Set, Tuple
+
+from .exceptions import CompileError
+
+
+class VariableType(Enum):
+    DATA = 1
+    PARAM = 2
+    ASSIGNED_PARAM = 3
+
+
+@dataclass()
+class TableRecord:
```

Comment:
Actually question, if this is the symbol table, are the symbols the keys or the values?

What I've been saying is that the symbols are the pieces of the AST. In this case, there can be multiple symbols (for instance, parts of the AST in different statements) that point to the same values in the symbol table.

Reply:
Right. I'd say it would be more appropriate to call it a "Variable table" instead. The SymbolTable keeps a single record for each variable, that contains any information that's necessary to construct the base dataframes and other subscripts.

Reply:
So you can have multiple Data AST nodes in the expression tree named `y`, but they share the same record within the table.

Reply:
Nice nice, so symbols from the AST and then these are still Variables.